crates/holochain/src/conductor/interface/websocket.rs

```diff
@@ -181,11 +191,89 @@ async fn recv_incoming_admin_msgs<A: InterfaceApi>(api: A, rx_from_iface: Websoc
         .await;
 }
 
+/// Takes an open connection and waits for an authentication message to complete the connection
+/// registration. If the connection is not authenticated within 10s or any other content is sent
+/// then the connection is dropped.
+///
+/// If the authentication succeeds then message handling tasks are spawned to handle normal
+/// communication with the client.
+fn authenticate_incoming_app_connection<A: InterfaceApi<Auth = AppAuthentication>>(
+    task_list: TaskListInner,
+    api: A,
+    mut rx_from_iface: WebsocketReceiver,
+    rx_from_cell: broadcast::Receiver<Signal>,
+    tx_to_iface: WebsocketSender,
+    installed_app_id: Option<InstalledAppId>,
+    port: u16,
+) {
+    let join_handle = tokio::task::spawn({
+        let task_list = task_list.clone();
+        async move {
+            let auth_payload_result = tokio::time::timeout(std::time::Duration::from_secs(10), async {
+                if let Ok(msg) = rx_from_iface.recv::<A::ApiRequest>().await {
+                    return match msg {
+                        ReceiveMessage::Authenticate(auth_payload) => {
+                            Ok(auth_payload)
+                        }
+                        _ => {
+                            warn!("Connection to Holochain app port {port} tried to send a message before authenticating. Dropping connection.");
+                            Err(())
+                        }
+                    }
+                }
+
+                warn!("Could not receive authentication message, the client either disconnected or sent a message that didn't decode to an authentication request. Dropping connection.");
+                Err(())
+            }).await;
+
+            match auth_payload_result {
+                Err(_) => {
+                    warn!("Connection to Holochain app port {port} timed out while waiting authenticating. Dropping connection.");
+                }
+                Ok(Err(_)) => {
+                    // Already logged, continue to drop connection
+                }
+                Ok(Ok(auth_payload)) => {
+                    let payload: AppAuthenticationRequest = SerializedBytes::from(
+                        holochain_serialized_bytes::UnsafeBytes::from(auth_payload),
+                    )
+                    .try_into()
+                    .unwrap();
+                    match api
+                        .auth(AppAuthentication {
+                            token: payload.token,
+                            installed_app_id,
+                        })
+                        .await
+                    {
+                        Ok(installed_app_id) => {
```

Comment:
Do we need to validate the installed_app_id against anything?

Reply:
This `installed_app_id` is the result of authenticating, so it's the one that was specified when the token was created. This is a dead-end at this point and in the next PR the requests and signals on the authenticated connection are tied to this `installed_app_id`.